carrier: merge carried item logic

lahm86 opened the pull request:
Resolves #2952.
Resolves #2954.
Resolves #2956.
Resolves #2957.

#### Checklist

- [x] I have read the [coding conventions](https://github.com/LostArtefacts/TRX/blob/master/CONTRIBUTING.md#coding-conventions)
- [x] I have added a changelog entry about what my pull request accomplishes, or it is an internal change
- [x] I have added a readme entry about my new feature or OG bug fix, or it is a different change

#### Description

This moves TR1's carrier logic to TRX and updates TR2 to use the same approach. Game flow defined drops remain a TR1-only feature, as these are required for the OG setup only, and the TR2 approach we had introduced previously is already compatible with TR2 itself. All enemies in TR2 can now drop, but the two eel types remain excluded as they can't be killed. Falling items now animate in 60 FPS as well.

I'll share a test level, but checking OG in both games would be good too. Checking current TR2X saves and legacy saves would be ideal as well.

lahm86 committed:
carrier: port carrier module to TRX

This ports the carrier module to TRX, although support remains only
with TR1 currently. TR2 will be transitioned.

lahm86 committed:
carrier: merge TR2 carried items

This ports TR2 to use the same carrier/drop logic as TR1. Falling items
are supported, and support has been added for water creatures, the
dragon and skidoo drivers to drop items. Eels will continue to not
pickup items.

Resolves #2952.

lahm86 committed:
carrier: fix dropped item interpolation

This ensures that items that are falling are marked as active so to
ensure they interpolate in 60 FPS.

Resolves #2954.

walkawayy reviewed (approved):
Looks great!

aredfan commented:
I noticed that `/kill all` causes the game to crash when using the tr1 test level, it only happens if tr2 item drops are enabled.

Although not a regression introduced in this PR, I thought it's worth mentioning.

aredfan commented:
- [x] I found just one bug - the dragon seem to drop the pickup prematurely with these steps.

1. launch the tr2 test level
2. `/save 1`
3. `/load 1`
4. activate the dragon **edit:** nvm you can see the pickup in bartoli's body, there's no need to activate the dragon

____________________

Oops I spoke too soon. I believe I have found another bug.

- [x] The `/kill` command seem to be buggy as Lara have to get really close to the dragon for it to be effective.
- [x] The `/kill` command also destroys the black skidoo, previously it would only kill the rider.
- [x] In the tr1 test level, if you save and reload, the mummy that stares at Lara drops the pickup prematurely. This bug is also present in the latest dev snapshot.

lahm86 committed:
carrier: fix Qualopec mummy early item spawn

This resolves the Qualopec mummy dropping items after save/load.

Resolves #2956.

lahm86 commented:
Thanks, @aredfan. The `/kill` commands should now be fixed, along with the mummy dropping prematurely. Still looking into the dragon early drop problem and the crash (also mummy-related, but having no luck so far).
I have a feeling the early drop for the mummy has been a problem since we introduced the TR2 style drops in 4.6. If you have time, could you double check if that's the case (just for the changelog)?

aredfan commented:
> Thanks, @aredfan. The `/kill` commands should now be fixed, along with the mummy dropping prematurely. Still looking into the dragon early drop problem and the crash (also mummy-related, but having no luck so far). I have a feeling the early drop for the mummy has been a problem since we introduced the TR2 style drops in 4.6. If you have time, could you double check if that's the case (just for the changelog)?

Anytime. 👍

I can confirm the mummy's early drop bug is a regression since 4.6, just checked.

lahm86 commented:
Thanks, @aredfan. Updated the changelog for the mummy.
The dragon early spawn should also be fixed too.
I'll look again at the mummy crash.

lahm86 committed:
tr1/lara/cheat: avoid disabling mummy AI

The Qualopec mummy uses its data field to store a single short for its
head rotation. When Lara cheats to kill the mummy, LOT is called to
disable the slot, but this can lead to memory corruption as it expects
a regular CREATURE pointer. For drop items, this was resulting in the
carried item's spawn number to be set to -1. The carrier module would
try to re-spawn the item, but the rest of the data was in a corrupt
state, and therefore either a crash or strange outcomes would result.

Resolves #2957.

lahm86 commented:
Fixed 😌 That was a well-hidden problem, explanation is in the commit message.

aredfan reviewed (approved):
> Fixed 😌 That was a well-hidden problem, explanation is in the commit message.

Well done and thank you. Everything LGTM 👍

walkawayy reviewed (approved):
Nice catches @aredfan
